GcExtensions.md

```diff
@@ -0,0 +1,353 @@
+  that determines whether a machine word is a pointer to an object,
+  including to its interior.
+
+To demonstrate the applicability and viability of these mechanisms, we
+have fully integrated Julia with the GAP computer algebra system, to the
+point that GAP's regular garbage collector is completely replaced with
+Julia's and that the lifetime of all GAP objects is entirely managed by
+Julia. We also implemented a self-contained test program using these
+mechanisms, and integrated it into the Julia test suite.
+
+The proposed implementation should not incur measurable overhead for
+Julia itself, as it only exposes additional functionality that is unused
+by Julia code, plus functionality hooks that are designed to only incur
+a few clock cycles of overhead per garbage collection. More specific
+discussion of overhead can be found accompanying the descriptions of
+these hooks.
+
+An implementation of this proposal can be found on GitHub under
+https://github.com/rbehrends/julia (branch `rb/gc-extensions`). See
+the example in the `test/gcext` subdirectory for an example of using
+this API.
+
+An implementation of GAP that uses the Julia GC in lieu of its native
+GC can likewise be found on GitHub at https://github.com/rbehrends/gap
+(branch `alt-gc`). This version of GAP can be built with:
+
+    ./autogen.sh
+    ./configure --with-gc=julia --with-julia=/path/to/julia/usr
+    make
+
+## Callbacks
+
+In order to allow foreign code to have access to necessary functionality
+in the garbage collector, we allow foreign code to register callbacks for
+certain GC events. We provide for six types of callbacks:
+
+* `pre_gc` (beginning of garbage collection).
+* `post_gc` (end of garbage collection).
+* `root_scanner` (before scanning GC roots).
+* `task_scanner` (when scanning Julia tasks).
+* `notify_external_alloc` (when an external object is allocated).
+* `notify_external_free` (when an external object is freed).
+
+With external objects, we refer to what in the current implementation are
+called `bigval_t` objects. These are allocated using the system's memory
+allocator rather than using Julia's external allocator. In order to not
+expose the implementation detail, we talk about "internal" and "external"
+objects rather than objects that are allocated as part of Julia's object
+pool or through system routines, respectively.
+
+For each type of callback, there is a corresponding function pointer type.
+Registering and deregistering callbacks occurs via macros.
+
+```
+typedef void (*jl_gc_cb_pre_gc_t)(int full);
+typedef void (*jl_gc_cb_post_gc_t)(int full);
+typedef void (*jl_gc_cb_root_scanner_t)(int full);
+typedef void (*jl_gc_cb_task_scanner_t)(jl_task_t *task, int full);
+typedef void (*jl_gc_cb_external_alloc_t)(void *addr, size_t size);
+typedef void (*jl_gc_cb_external_free_t)(void *addr);
+
+#define jl_gc_register_callback(cb, func)
+#define jl_gc_deregister_callback(cb, func)
```

Comment:
Implementation notes. It seems that there's no reason these should be macros. What they do can be implemented with functions without any problems. The GC code used to be full of such macros saving negligible typing and making the code very hard to reason about/search. These should be replaced with normal functions.

Reply:
Unfortunately, this cannot be done with functions in a typesafe fashion, as the type of the second argument depends on the value of the first. This is an attempt to work around the lack of generics in C.

That said, I'd be very much open for alternative solutions.

Reply:
Specialized registration function for each or use `void*` as function pointer argument type.

Reply:
Also, the worst part of this is making thing much harder to search, i.e. with the symbol concatenation. A macro with explicit spelling is also OK (same goes for the few new macros in `gc.c`).

Reply:
Using `void *` would loose the type safety -- but I'd be fine with that, too (it's just another very small convenience for detecting very basic errors, but hardly important). And of course you are right, we could replace them by 2*6 custom (de)register functions, one for each hook -- I think the main annoying bit about that is the code duplication -- but if that's what you guys prefer, we could do it. We'll wait for some more feedback (perhaps some other Julia devs also have comments) and then look into revising the proposal. Thanks for your comments so far!

Reply:
Specialized registration functions make the interface very cumbersome, but it could be done. `void *` is not an option, as function pointers cannot be safely cast to `void *` (they can be safely cast to an arbitrary function pointer type and back, but that makes client code very verbose). And in either case, we would lose type safety. The way to go in a typesafe fashion would be to have separate registration/deregistration functions for all callbacks.